# Function Parameters

## Argument vs parameters

In [2]:
def my_func(a,b):
    return a*b

V tomto kontextu a a b jsou parametry funkce. Pokud ale funkci zavolam tzn my_func(x,y) jsou argumenty funkce.

- v podstate parametr je defaultni hodnota ktera se stane argumentem az v akci :]
- argument je az u funkce co neco prebira
- dulezite je brat v potaz ze nase parametry a pote argumenty - tak funkce obdrzi reference kde najde potrebne hodnoty

## Positional and Keyword Arguments

### Teorie

- nejcastejsim stylem zadavanim argumentu je skrze poradi, pokud neurcim a zapisi pouze hodnoty tak si to funkce prebere sama
- mohu specifikovat defaultni hodnoty, cimz znich vytvarim dobrovolne argumenty - pokud nespecifikuji vezme defaultni hodnotu
- my_func(a, b = 100, c) - A JAK zapisi ted? Nejde, musel bych definovat i defaultne hodnotu c - JINAK RECENO - DEFAULTNI HODNOTY VZDY NA KONCI
- Samozejme to lze resit ale skrze keyword argumenty - tzn definovat CO = CO :], tzn. keyword neboli named argument
- Jakmile pouzivam keyword a named argumenty nepotrebuji dodrzovat poradi, samozrejme jakmile nejaky argument pojmenuji musim pojmenovat i dalsi - logicke ..porad je to jen komunikace, my_func(1,b=2,c=3) - funguje, plati ze do bodu dokud nezacnu pouzivat keyword ..takze po prvnim pojmenovani musim pojmenovvat dalsi

### Code

In [4]:
def my_func(a,b,c):
    print (f'a={a}, b={b}, c={c}')

In [6]:
my_func(1,2,3)

a=1, b=2, c=3


In [7]:
def my_func(a,b=2,c):
    print (f'a={a}, b={b}, c={c}')
#jasne a nefunkcni

SyntaxError: non-default argument follows default argument (<ipython-input-7-ccd5d261afbc>, line 1)

In [9]:
def my_func(a=10,b=20,c=30):
    print (f'a={a}, b={b}, c={c}')

In [11]:
my_func() #vse defaultni nemusim zadavat nic 

a=10, b=20, c=30


In [12]:
def my_func(a,b=2,c):
    print (f'a={a}, b={b}, c={c}')

In [15]:
my_func(b = 4, c = 8, a = 4) 

a=4, b=4, c=8


Zde pouzivam kezword argumenty. To uz je ciste na me. Poradi si muzu urcit. Jinak netreba resit vsechno je v teorii.

## Unpacking Iterables

### Teorie P1

- 1,2,3 je tuple
- Jinymi slovy - zavorky netvori tuple
- Pouze pokud chci vytvorit prazdny tuple pote staci ()
- Nebo tuple() construktor

In [17]:
1,2,3 #tuple bez zavorek

(1, 2, 3)

Packed Values

- v podstate vse co lze iterovat je packed values..Listy, stringy, dictionaries apod..
- unpackingem tak priradim hodnoty
- a,b,c = [1,2,3] --> a1 b2 c3, poradi je postupne
- a,b, c = 10, 20 , 'hello' --> v podstate 2 tupply co se spojujou 
- a,b, c = 'XYZ' aX bY cZ

Swapping

In [19]:
# tmp = a
# a = b
# b = tmp
#Tahle logika je jasna ulozim si a do prechodne, prehodim a pote vratim

Mame zde ale swapping v badass stylu

- a,b = b,a
- jede z pravda do leva v tomto pripade takze si vlevo vezme adresz pro a a b a pote je priradi pod nove nezname 

Unpacking Sets and Dictionaries

Dictionaries and sets jsou unordered - ikdyz mozna uz to tak neplati? 

In [34]:
d = {'key1': 1,'key2': 2,'key3': 3}

In [35]:
a,b,c = d

In [36]:
a,b,c

('key1', 'key2', 'key3')

Zkusime tedy set

In [37]:
s = {'a','b','c','d'}

In [38]:
a,b,c,d = s

In [39]:
a,b,c,d

('a', 'c', 'd', 'b')

Takze dictionaries jsou razeny nejspise? Ale sets jsou uz vice mimo.

### Coding P1

In [41]:
s = 1,2,3
type(s) #jen mensi ukazka :]

tuple

In [42]:
a = (1)
type(a) #na tohle pozor

int

In [43]:
a = (1,)
type(a) #carka to je ten tvurci element

tuple

In [44]:
a = 100,
type(a)

tuple

In [45]:
a = ()
type(a) #a takto prazdny tuple

tuple

In [46]:
a, b, c = 10, {1,2}, ['a','b']
c,b,a #mensi ukazka 

(['a', 'b'], {1, 2}, 10)

In [47]:
a, b = 10, 20
a, b = b, a
a, b #Nas perfektni swapping

(20, 10)

In [48]:
a,b,c = 'XYZ'
c,b,a

('Z', 'Y', 'X')

In [50]:
d = {'key1': 1,'key2': 2,'key3': 3}
a,b,c = d.values()
a,b,c
#tohle je mensi ukazka ze lze unpuckovat i takto

(1, 2, 3)

In [51]:
d = {'key1': 1,'key2': 2,'key3': 3}
a,b,c = d.items()
a,b,c

(('key1', 1), ('key2', 2), ('key3', 3))

Pokud unpackuji dicitonarlne dostanu klice, pokud pres values obdrzime hodnoty, pres items dostaneme tuple

In [52]:
d = {'key1': 1,'key2': 2,'key3': 3}
(a,b),(c,d),(e,f)= d.items()
f #tada kompletni tuple unpacking

3

### Unpacking teorie P2

OPERATOR *

- nekdy chceme unpackovat pouze cast a ne vse a nebo cast a vse apod. na to mame *
- l = [1,2,3,4,5,6]
- a, b = l[0], l[1:] --> prvni a zbytek a pouze jen u sekvence co lze slicovat lepsi je takto
- a, *b = l --> v tomto pripade prvni a zbytek
- a, b, *c --> tez funkcni tzn.. *znak je v podstate zbytek
- a, b*, c ---> PRVNI - VSE KROME - POSLEDNI :]
- a*,b* nelze zapsat 
- tez mohu diky tomuto rozbalovat vice veci naraz
- l = [a*,b*] -- a = [1,2,3], b = [4,5,6], l = [1,2,3,4,5,6]
- * operator funguje i u setu a dictionary - ale zde mame vetsi problem s presnosti..tzn. nevime vysledek (u dicionary mozna ano)
- mohu ale vyuzit stejne jako u listu tzn na prave strane unpackovat veci do veci 

OPERATOR **

- ** nelze pouzivat na leve strane
- ale mohu je pouzit napriklad pro merge dictionaries, posledni hodnoty prepisuji, vyhoda i nevyhoda - skvele pro prepisovani hodnot

NESTED UNPACKING

- l - [1,2[3,4]], a, b, c = l --> a = 1, b =, c = [3,4]
- badass unpacking je a, b, (c, d)

a, *b, (c, *d), = [1,2,3,'python'], a = 1, b = [2,3], c = 'p', d = ['y't'h'o'n]

To je perlicka na zaver. Pomoci slicingu by to slo o neco hure. Nicmene hlavni bod je jasny, lze pouzit vickrat * pokud je to pro sve celky

### Coding P2

Zaklady unpackingu jiz znam takze tolik neresim a projedu rychleji.

In [83]:
l = [1,2,3,4,5,6]

In [86]:
a, b  = l[0], l[1:]
a,b

(1, [2, 3, 4, 5, 6])

In [88]:
a, *b = l
b

[2, 3, 4, 5, 6]

In [93]:
s = 'python'
a, *b, c = s
b

['y', 't', 'h', 'o']

In [94]:
a, b, c, d = s[0], s[1], s[2:-1], s[-1]
a,b,c,d

('p', 'y', 'tho', 'n')

In [95]:
*s, = s

In [96]:
s

['p', 'y', 't', 'h', 'o', 'n']

Tohle je moc pekny zpusob jak vytvaret list :]

In [98]:
l1, l2 = [1,2,3],[4,5,6]

In [99]:
l = [*l1, *l2]
l #moc pekny zpusob pro spojeni dvou listu

[1, 2, 3, 4, 5, 6]

In [101]:
s1 = {9,8,7}
[*l1, *s1] #set uz nedrzi poradi

[1, 2, 3, 8, 9, 7]

In [103]:
s1, s2 = 'abc', 'cde'
s3 = {*s1,*s2} #je v podstate jedno kam unpackuji
s3

{'a', 'b', 'c', 'd', 'e'}

In [104]:
*c, = s3
c

['b', 'c', 'd', 'e', 'a']

Propojeni setu

In [109]:
s1 = {1,2,3,4,5}
s2 = {5,6,7,8,9}

In [107]:
s1+s2 #takto nejde

TypeError: unsupported operand type(s) for +: 'set' and 'set'

In [110]:
{*s1, *s2} #takto jde - prepisuji se shody

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [111]:
s1.union(s2) #lze tez ale proc..takto to jde lepe muzu si i vybirat

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [112]:
d1 = {'key1' :1,'key2' :2}
d2 = {'key2' :3,'key4' :4}

In [113]:
{*d1,*d2} #tohle zrovna nechci 

{'key1', 'key2', 'key4'}

In [115]:
{**d1,**d2} # tadaaa :]
#key 2 nam to prepsal klasika

{'key1': 1, 'key2': 3, 'key4': 4}

V podstate cil je mergovat v poradi v jakem potrebuji prepisovat - aktualnejsi informace az pote.

In [116]:
a, *b,(c,*d)  = [1,2,3,4,'python']

In [119]:
a, b, c, d

(1, [2, 3, 4], 'p', ['y', 't', 'h', 'o', 'n'])

Nested unpacking, tez jsem pomerne obeznamen.

## * ARGS

### Args teorie

- v podstate args parametr funguje na stejne logice jako unpacking s hvezdickou
- *args je bezna konvence - podle me lepsi normal jmeno pokud plni nejaky presny ucel
- *args vycerpa vsechny parametry - tzn. musi byt vzdy vedena jako posledni

- def func1(a,b,c), l = [10,20,30]
- func1(l) -> nefunguje
- func(*l) -> funguje

### Code

In [ ]:
a, b, *c = 10,20, 'a', 'b'

In [121]:
def func1(a, b, *c):
    print(a)
    print(b)
    print(c)

In [122]:
func1(10,20,1,2,3,4,5,6)

10
20
(1, 2, 3, 4, 5, 6)


V podstate si args argument prevzal vice hodnot a vytiskl jako tuple. Takze vse ok.

In [128]:
#jednoducha funkce pro prumer
def avg(*args):
    count = len(args)
    total = sum(args)
    return total / count
#neumi resit 0

In [130]:
avg(10,20,4,5,6)

9.0

In [131]:
def avg(*args):
    count = len(args)
    total = sum(args)
    if count == 0:
        return 0
    else:
        return total / count
#slabsi ale funkcni reseni    

In [132]:
avg(0)

0.0

In [133]:
def avg(*args):
    count = len(args)
    total = sum(args)
    return count and total/count
    #nase short circuit ficura

In [134]:
avg(0)

0.0

In [135]:
#obejdu 0 tim ze donutim urcit aspon jeden parametr
def avg(a, *args):
    count = len(args) +1
    total = sum(args) +a
    return total / count

In [137]:
def func1(a, b, c):
    print(a)
    print(b)
    print(c)

In [138]:
l = [10,20,30]

In [141]:
func1(*l) #tohle je moc pekne :]

10
20
30


Unpacknu hodnoty a prebere si to krasne.

In [142]:
l = [10,20,30,40,50]

In [143]:
def func1(a, b, c, *args):
    print(a)
    print(b)
    print(c)
    print(args)

In [144]:
func1(*l)

10
20
30
(40, 50)


A tohle je jeste hezci unpacking.

## Keyword arguments

### Teorie + coding

- povinny keyword argument lze nadefinovat skrze *args - po *args musi byt kezword argument nebo nam nebude rovnice fungovat

In [145]:
def func(a,b,c):
    print(a,b,c)

In [146]:
func(1,2,3)

1 2 3


In [147]:
func(1, c=3, b=2)

1 2 3


In [150]:
def func1(a, b, *args):
    print(a,b, args)

In [151]:
func1(1,2,3,4) #jasne args vyuzije zbytek parametru

1 2 (3, 4)


In [152]:
def func1(a, b, *args, d):
    print(a,b, args, d)

In [161]:
def func1(*args, d):
    print(args, d)

In [162]:
func1(1,2,3, d="a")

(1, 2, 3) a


In [ ]:
#můžu takto zavolat

In [165]:
func1(d="a")

() a


In [ ]:
#tak a pokud chci abych nemohl zadávat žádní positional argumenty můžu takto:

In [166]:
def func(*, d):
    print(d)

In [168]:
func(d=100) #v tomtp případě musím dodat povinně parametr

100


In [169]:
def func(a,b,*,d):
    print(a,b,d)
#v podsattě tím říkám jedno - můžeš zadat dvě volné hodnoty
#a třetí musí být keyword

In [170]:
func(1,2, d=4) #takto mohu

1 2 4


In [171]:
def func(a, b=2, *args, d):
    print(a,b,args,d)

In [173]:
func(1,5,3,4,6, d="a")
#trochu bordel jak vím kdy chci zadat b a kdy ne?

1 5 (3, 4, 6) a


In [174]:
def func(a,b =20, *args, d=0, e):
    print(a,b,args,d,e)

In [176]:
func(5,4,3,2,1, e = "all engines running")
#v podstate jsem urcil ze e - musí být povinně zadáno, d je defaultn a je ignorovano pokud jej nebudu definovat

5 4 (3, 2, 1) 0 all engines running


In [178]:
func( 0, 600, d="good morning", e="python")
#zde jsem vynechal args - takze zde neni

0 600 () good morning python


In [180]:
func(11, "m/s", 24, "mph", d="unladen", e="swallow")
#nyní mám dvě dané, poté args a poté určené - obecně je to docela bordel

11 m/s (24, 'mph') unladen swallow


Tohle je obecně docela ghulas. Takže je nezbytné to dát řádně dokupy.

### **kwargs

U funkce je to jasné, * pro positional argumenty a ** pro keyword argumenty.

- po ** kwarfs už nesmí být žádné parametry
- def func(*, d, **kwargs): --> rikáme tedy ze nemáme žádné positional arguemnty, a že náš d je povinný k zadání, a poté máme zbytek ve formě dictionary
- def func(**kwargs): --> zadavama proste dt
- def func(*args,**kwargs): --> nejsou povinne, ale mohu takto func(4,5,a=50,b=60)
- v podstatě je sranda že já si pamratery u kwargs definuji až potom

In [182]:
def func(**others):
    print(others)

In [183]:
func(a=1, b=2, c=3)

{'a': 1, 'b': 2, 'c': 3}


In [184]:
def func(*args, **kwargs):
    print(args)
    print(kwargs)

In [185]:
func(1, 2, x=100, y = 200)

(1, 2)
{'x': 100, 'y': 200}


In [187]:
def func(a,b, *, **kwargs):
    print(a)
    print(b)
    print(kwargs)
    #nebude fungovat 

SyntaxError: named arguments must follow bare * (<ipython-input-187-f3eb126a4eb1>, line 1)

Pouze pokud bych měl za * named argument.

In [191]:
def func(a,b, **kwargs):
    print(a)
    print(b)
    print(kwargs)


In [192]:
func(1,2,x=100,y=200)

1
2
{'x': 100, 'y': 200}


V podstatě kwargs jsou keyword arguemnty.

### Putting it all together teorie

- a,b c = 10, args, hvezdicka , kw1, kw2 = 100, kwargs
- a,b musí být zadány, c má defaultní hodnotu nemusí být
- po args sbíráme další arguemnty
- hvezdicka ukončuje všechny postiional arguemnty
- poté následují keyword only arguemnty - znovu mohou mít defaultní hodnotu
- tzn. kw1 je povinný musíme zadat, kw2 ne díky defaultní hodnotě
- kwargs posbírá dodatečný kw hodnoty
- abych mohl přejít na kw arguemnty musím nejdříve vyčerpat positional 
- tedy přimu vše co tam je pomocí args a nebo je vyčerpám hvezdickou

- def func(a, b = 10, hvezdicka, kw1, kw2=100)
- def func(a, b, args, kw1, kw2=100)
- defaultní b u * může být ale pokud mám *agrs
- nemůžu ukončovat positional pomocí *args defaultní hodnotou zmatu python
- def func(a, b=10, *, kw1, kw2=100, kwargs)
- def func(args)
- def func(kwargs)
- def func(args, kwargs)
- pouze příklady jak lze definovat funkce

- Positional arguments
- mohou mít ale nemusí defaultní hodnoty
- args posbírá a vyčerpá zbývající positional arguemnty

- Keyword only argument
- Musí zadat klíč i hodnotu, a vždy po vyčerpání positional
- kwargs stejné jako pro args jen pro key-value argumenty
- hvezdicka indikuji konec positional argumentů


### Putting it all together Code

In [1]:
def func(a,b, *args):
    print(a,b,args)

In [2]:
func(1,2,"x","y","z")

1 2 ('x', 'y', 'z')


In [3]:
func(a=1,b=2,"x","y","z")

SyntaxError: positional argument follows keyword argument (<ipython-input-3-dbc1c2f9fdd0>, line 1)

Jakmile mám args nemohu tam nacpat zápis takto.

In [9]:
def func(a,b=2,c=3,*args):
    print(a,b,c,args)

In [5]:
func(1,c=5,"x","y")

SyntaxError: positional argument follows keyword argument (<ipython-input-5-912dec912bcc>, line 1)

V podstatě jakmile nastavím defaultní hodnoty a nastavím *args ztrácím možnost použít defaultní hodnoty.

In [11]:
def func(a,b=2,*args, c=3, d):
    print(a,b,args,c,d)

In [12]:
func(10,20,"x","y","z", c = 4, d=1)

10 20 ('x', 'y', 'z') 4 1


In [13]:
func(1, "x","y","z", b=4, d = 10)

TypeError: func() got multiple values for argument 'b'

Zde si myslí že x,y,z je b též.

In [15]:
def func(a,b,*args, c=10, d=20, **kwargs):
    print(a,b,args,c,d,kwargs)

In [16]:
func(1,2,"x","y","z", c=100, d=200, x=0.1, y=0.2)

1 2 ('x', 'y', 'z') 100 200 {'x': 0.1, 'y': 0.2}


Kwargs je docela stylovka prostě funkce s parametrem a hodnotou zároveň.

Menší ukázka proč je vhodné chápat argumenty u funkcí a že hodně častých funkcí má v sobě mnoho defaultních funkcí

In [28]:
print(1,2,3)
print(4,5,6)

1 2 3
4 5 6


In [30]:
print(1,2,3, sep = "-")
print(4,5,6, sep="-")

1-2-3
4-5-6


In [31]:
print(4,5,6, sep="-", end="***")

4-5-6***

In [32]:
def calc_hi_lo_avg(*args, log_to_console=False):
    hi = int(bool(args)) and max(args)
    #args může být prázdný jak když nic nezadají? Jak zadat 0
    #provedu luxusní bol zápis args and :D ..
    #vezmu args hodnotu v bool - zda je True nebo False a to převedu na int
    #tzn. pokud nebudu mít hodnotu - mám false, a tím pádem mám bool hodnotu 0 a tudíž mám průměr 0
    if len(args) == 0:
        lo = 0
    else:
        lo = min(args)
    #ekvivaleltní low hodnota pro hi nahoře - jen jiný zápis
    #pro mě docela mindfuck
    #lo = min(args) if len(args) > 0 else 0 
    avg = ( hi + lo) / 2
    if log_to_console:
        print(f"high={hi}, low={lo}, avg={avg}")
    return avg

In [33]:
avg = calc_hi_lo_avg(1,2,3,4,5, log_to_console=True)

high=5, low=1, avg=3.0


Jasné moje funkce vrací jen average pokud neurčím defaultně jinak, poté mi vyhodí všechny hodnoty, tj moc pěkné použití.

### Application - A Simple Function Timer

Budeme chtít napsat funkci, která umí měřit jak dlouho to dané funkci trvá.

In [34]:
import time

In [35]:
def time_it(fn, *args, **kwargs):
    print(args, kwargs)

In [36]:
time_it(print, 1, 2, 3, sep = " - ", end = " **** ")

(1, 2, 3) {'sep': ' - ', 'end': ' **** '}


V podstatě mám funkci, která nám něco posbírá a jen vyplivne.

In [37]:
def time_it(fn, *args, **kwargs):
    fn(args, kwargs)

In [38]:
time_it(print, 1, 2, 3, sep = " - ", end = " **** ")

(1, 2, 3) {'sep': ' - ', 'end': ' **** '}


Proč to nefunguje? 

- v prvním případě to dělá co chceme jen tiskneme a nic nevoláme
- v druhém ale chceme volat funkci s parametry args a kwargs
- jenže naše volání funkce lze přepsat takto
- fn(args,kwargs) = print(args, kwargs)
- ALE FUNKCE FUNGUJE NA * a ** - tzn musíme zadat i zde

In [43]:
def time_it(fn, *args, **kwargs):
    fn(*args, **kwargs)

In [44]:
time_it(print, 1, 2, 3, sep = " - ", end = " **** ")

1 - 2 - 3 **** 

A nyní nám to funguje :). Funkce co volá funkci nic víc.

In [49]:
def time_it(fn, *args,rep=1, **kwargs):
    for i in range(rep):
        fn(*args, **kwargs)
    #nově dodávám rep abych mohl zadat kolikrát chci aby se funkce opakovala

In [50]:
time_it(print, 1, 2, 3, sep = " - ", end = " **** ", rep = 5)

1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 

Naše krásná funkce co volá funkci a ještě jí dodává parametry pff..

In [53]:
def time_it(fn, *args,rep=1, **kwargs):
    start = time.perf_counter()
    for i in range(rep):
        fn(*args, **kwargs)
    end = time.perf_counter()
    return (end - start) / rep

In [54]:
time_it(print, 1, 2, 3, sep = " - ", end = " **** ", rep = 5)

1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 1 - 2 - 3 **** 

0.0005012600000554812

Nyní naše skvělá funkce ještě měří čas.

Nyní trochu funkce co zaberou delší čas

In [66]:
def computer_power_1(n, *, start=1, end):
    #using a for loop
    results = []
    for i in range(start, end):
        results.append(n**i)
    return results
#nic extra prostě jen obrždí n - začátek a konec
#a dává na druhou podle i a háže do listu

In [59]:
computer_power_1(2, end=5)

[2, 4, 8, 16]

In [65]:
def computer_power_2(n, *, start=1, end):
    #using a list comprehension
    return [n**i for i in range(start, end)]
    #jen jiný zápis

In [67]:
computer_power_2(2, end=5)

[2, 4, 8, 16]

In [69]:
def computer_power_3(n, *, start=1, end):
    #using generators expression
    return (n**i for i in range(start, end))
#pomocí generátoru tedy

In [73]:
list(computer_power_3(2, end = 5))
#zajímavé generátor má změnu oropoti comprehension že je v kulatých závorkách
#a dále, že odkazuje na objekt takže abych jej vytiskl musím jej převést na list při volání

[2, 4, 8, 16]

Teď je budeme časovat a zjištovat která je nejryhclejší.

In [79]:
time_it(computer_power_1, 2, start=0, end=20000, rep=5)

0.5064703000000008

In [80]:
time_it(computer_power_2, 2, start=0, end=20000, rep=5)

0.49456472000001667

In [81]:
time_it(computer_power_3, 2, start=0, end=20000, rep=5)

3.899999956047395e-06

Tohle je moc pěkná funkce obecně pro časování. Funkce poslední však jen vytváří generátor nevrací stejná data proto tak rychle.

Musel bych vracet list generátoru

In [83]:
def computer_power_4(n, *, start=1, end):
    #using generators expression
    return list((n**i for i in range(start, end)))

In [84]:
time_it(computer_power_4, 2, start=0, end=20000, rep=5)

0.4098143000001073

Jsem o chlup rychlejší ale už né tak moc.

### Parameter Defaults - Beware!

Řešíme problém, když máme parametry které se načtou do paměti a při volání je nepřegenrují ale pouze zavolají. Typickí příklad je s časem, kdy máme stále stejný čas. Proto je vhodné dávat defaultně None hodnoty pro tyto druhy parametrů a pokud nejsou non tak je volat.

In [85]:
from datetime import datetime

In [88]:
print(datetime.utcnow())

2020-05-05 17:13:00.797579


In [90]:
print(datetime.utcnow()) #vidíme, že čas se mění

2020-05-05 17:13:16.059331


In [97]:
def log(msg, *, dt=datetime.utcnow()):
    print(f"{dt} and {msg}")

In [98]:
log("message 1", dt="2001-01-01 ...")

2001-01-01 ... and message 1


Jasná funkce a něco plodíme, čas jsem si zadal svůj takže je vše ok.

In [99]:
log("message 2")

2020-05-05 17:14:55.640335 and message 2


Nyní vidíme, že když nezadáme datetime použijeme funkci.

In [100]:
log("message 3")

2020-05-05 17:14:55.640335 and message 3


#### A TO JE NÁŠ PROBLÉM - ČAS SE NEZMĚNIL

2020-05-05 17:14:55.64033, 2020-05-05 17:14:55.64033

Defaultní hodnota totiž byla vytvořena v čtverci IN[97] - pokud bych jí znovu spustil a přehrál funkce, mám zase 2 stejné časy, jen jiné.

Zkusíme to tedy vyřešit skrze None hodnotu

In [102]:
def log(msg, *, dt=None):
    if not dt:
    #pokud dt je falsy - pokud jsem jej nezadal tak načtu dle funkce
        dt = datetime.utcnow()
    print(f"{dt} and {msg}")

In [105]:
log("message 4")

2020-05-05 17:18:32.764441 and message 4


In [106]:
log("message 4")

2020-05-05 17:18:33.354833 and message 4


A nyní krásně vidím, že mám vše v pořádku :). Ale lepší zápis.

In [108]:
def log(msg, *, dt=None):
    dt = dt or datetime.utcnow()
    # pokud dt má hodnotu použije dt 
    # nebo použije datetime.utcnow() :)
    print(f"{dt} and {msg}")

Teď trochu z jiného soudku

In [109]:
my_list = [1,2,3]

def func(a=my_list):
    print(a)

In [110]:
func()

[1, 2, 3]


In [111]:
func(["a","b"])

['a', 'b']


In [112]:
my_list.append(4)

In [113]:
my_list

[1, 2, 3, 4]

In [114]:
func()

[1, 2, 3, 4]


Jaký je zde problém? Mám ve funkci odkaz na objekt mimo. Pokud tedy nechci aby nedošlo ke změně je lepší použít tuple.

### Parameter Default Beware Again

In [116]:
def add_item(name, quantity, unit, grocery_list):
    grocery_list.append(f"{name}, {quantity}, {unit}")
    return grocery_list

In [117]:
store1, store2 = [], []

In [118]:
add_item("banana", 2, "units", store1)
add_item("milk", 1, "liter", store1)

['banana, 2, units', 'milk, 1, liter']

In [120]:
add_item("python", 1, "medium-rare", store2)

['python, 1, medium-rare']

Prostě mám jen funkci o jméně množství jednotce a listu a vkládám do lsitu nic víc

In [121]:
#nyní si ve funcki zakládám list sám, abych je nemusel definovat
def add_item(name, quantity, unit, grocery_list=[]):
    grocery_list.append(f"{name}, {quantity}, {unit}")
    return grocery_list

In [123]:
del store1
del store2

In [124]:
store1 = add_item("banana", 2, "units") #vytvoří nový list
add_item("milk", 1, "liter", store1)# je vytvořený tak jej použije

['banana, 2, units', 'milk, 1, liter']

Tohle je hodně zajímavé :D

In [125]:
store2 = add_item("python", 1, "medium-rare") #nový list store 2

In [126]:
print(store2)

['banana, 2, units', 'milk, 1, liter', 'python, 1, medium-rare']


In [127]:
store1 is store2

True

HMM nastal nám problém. Podle všeho naše supr tvorba listu není nijak přiřazená takže return glocery list se stackuje - jedno co si napíšu předtím.

Náš default [] není žádná extra tvorba listu ale má stejnou adresu a proto nám nastává problém.

Jak to pořešíme?

In [129]:
def add_item(name, quantity, unit, grocery_list=None):
    if not grocery_list:
        grocery_list = []
    #pokud neuvedeme přesěn list a máme None tak vytvoříme nový
    grocery_list.append(f"{name}, {quantity}, {unit}")
    return grocery_list

Nyní s None by naše fičura se store1 a store 2 fungovala.

In [130]:
store1 = add_item("banana", 2, "units") #vytvoří nový list
add_item("milk", 1, "liter", store1)# je vytvořený tak jej použije
store2 = add_item("python", 1, "medium-rare") #nový list store 2

In [131]:
print(store1)

['banana, 2, units', 'milk, 1, liter']


In [132]:
print(store2)

['python, 1, medium-rare']


In [133]:
store1 is store2

False

Nyní je vše v pořádku. Takže pozor na jaké úrovni vytváříme listy a obecně mutable objekty.

A NYNÍ NOVÝ SOUDEK

In [134]:
def factorial(n):
    if n < 1:
        return 1
    else:
        print(f"calculating{n}")
        return n* factorial(n-1)

In [135]:
factorial(3)

calculating3
calculating2
calculating1


6

Když bych se opravdu nudil mohu dodělat poslední funkci

OBECNĚ JE OPRAVDU NEZBYTNÉ TO VŠE ZNOVU ČÍST A STUDOVAT - JE TO DOCELA FUŠKA A JE NUTNÉ DOSTAT VŠE POD KŮŽI